[RFC] stm32h7: Enable support for optional/alternate clocks

erwango opened the pull request:
## TL;DR ##
STM32 clock_control driver update (limited to H7 for now) to provide a way to configure alternate clocks on peripherals that support it:
```
    clocks = <&rcc STM32_CLOCK_BUS_APB2 0x00001000>,  <- Existing, bus driven, "register" clock
             <&rcc STM32_SRC_PLL3_P SPI123_SEL(2)>;   <- Newly added alternate clock (PLL3_P output here)
```
This alternate clock source could be selected in peripheral driver thanks to new `clock_control_configure()` API:
```
	if (IS_ENABLED(STM32_SPI_OPT_CLOCK_SUPPORT) && (cfg->pclk_len > 1)) {
		if (clock_control_configure(DEVICE_DT_GET(STM32_CLOCK_CONTROL_NODE),
					    (clock_control_subsys_t) &cfg->pclken[1],
					    NULL) != 0) {
			LOG_ERR("Could not select alternate SPI source clock");
			return -EIO;
		}
	}
```

## Goal
Aim of this work is to enable the use of an optional clock in a device configuration.
This is particularly required in case of STM32H7 based platform on which device clock configuration is usually based on 2 different clocks (cf #41650)
- A register clock, required for device configuration, usually coming from APB/AHB bus.
- A kernel clock, which will be used to clock the hardware block (and hence is the one driving the device frequency), typically derived from an alternate clck source (PLL output, fixed clock, ..)

Fixes #41650

### Why it worked so far on H7
Device clock configuration used to be functional up to now since clock configuration was set so that APB clk and alternate device clk (the default reset one) were equivalent. 

### Secondary goal
Besides the initial goal, idea was to provide the basis to support equivalent functionalities on onther STM32 series, such as the possibility to provide an alternate clock to a peripheral. For instance, configuring LSE as the LPUART or LPTIM clock, or HSI48 as the USB clock.

## Update of clock_control API (03/14):
 
Reworked to make use of new `clock_control_configure()` API, cf https://github.com/zephyrproject-rtos/zephyr/pull/43790

## Implementation options: 

### Alternate clocks bindings
Additional/Alternate clocks could be fixed clocks (HSI, HSE, ..), PLL outputs (PLLX_P/Q/R), sysclock, ....
For all these clocks, I've decided to keep existing RCC device driver as the clock controller. As a consequence, the dt binding used to select a specific clock looks to:
```
	clocks = <&rcc STM32_CLOCK_BUS_APB2 0x00001000>,  <- Existing, bus driven, "register" clock
		 <&rcc STM32_SRC_PLL3_P SPI123_SEL(2)>;   <- Newly added alternate clock (PLL3_P output here)
```
Another way to support this would have been to set up each potential clock source as a clock controler on its own and use the following binding:
```
	clocks = <&rcc STM32_CLOCK_BUS_APB2 0x00001000>,  <- Existing, bus driven, "register" clock
		 <&pll3 pll_p_output SPI123_SEL(2)>;      <- Discarded binding option
```
I've discarded this second option because, to be consistent with the binding, this second option would require to instantiate each potential source clock as a device driver, each one implementing clock_control API. This would have had negative impact on code footprint, which could have been considered ok for STM32H7 series but problematic on "smaller" STM32 series.

### Clock mux bindings
Complete solution requires to find a way to configure clock mux (as per_ck).

First option is to add a dedicated node descibing  the clock mux. Its only property is to configure its clock input
```
&perck {
	clocks = <&rcc STM32_SRC_HSI_KER CKPER_SEL(0)>;
	status = "okay";
};
```
Works fine, main drawback is the cost (~220 bytes of flash)

A much simpler alternative was to add clock configuration directly to the clock consumer end node:
```
	clocks = <&rcc STM32_CLOCK_BUS_APB2 0x00001000>,
		 <&rcc STM32_SRC_HSI_KER CKPER_SEL(0)>,
		 <&rcc STM32_SRC_CKPER SPI123_SEL(4)>;
```
Though, this proposal has 2 main drawbacks:
- If board dts is poorly configured, 2 consumers of the same internal clock (per_ck here) could set conflicting configurations
- This adds complexity on the consumers driver implementation side that should be able to differentiate between all these clocks 

In the end the addition of a clock mux node device felt saner.

### Dependencies
This work is based on:
- ~~https://github.com/zephyrproject-rtos/zephyr/pull/41585, which contain the preliminary clock_control driver work~~ merged
- ~~New set of `/include/devicetree/clock.h` macros (cf https://github.com/zephyrproject-rtos/zephyr/pull/42326)~~ merged
- Dedicated tests provided in `tests/drivers/clock_control/stm32_clock_configuration/stm32h7_devices/`

### Limitations and next steps:
- Only STM32 SPI driver is made aware of this. All STM32 drivers should be updated to take this new clock config into account.
- Extend to other STM32 series. This kernel clock is generally not available on other STM32 series, but most of this work can be reused to support the configuration of alternate peripheral clocks (like HSI on LPUART). This extension would first require a clock_control driver clean up similar to #41585.
- It is not possible today to `clock_off` supplementary clocks as they potentially clocks multiple peripherals (or even the core itself). So, this most likely requires the use of power-domains

erwango committed:
tests/drivers/clock_control: stm32h7: Move tests under stm32h7_core

Before introducing a new test for peripheral clocks,
rename existing stm32h7 test section by stm32h7_core.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
tests/drivers/clock_control: stm32h7: Add test for devices clock cfg

Add 2 clocks tests around device clock configuration on stm32h7.
For now, 'spi1_pllq_2_d1ppre_4' test variant is failed, which
illustrates issue reported in #41650.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
dts/arm: stm32h7: Introduce a stm32h7 specific clock binding

This new binding allows to work on providing stm32h7 specific
alternate and complementary device clocks.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
drivers/clock_control: stm32h7: Change clock bus bindings values

Set bus binding values using registers offset values.
As a consequence update driver to take this into account
in clock_on and clock_off functions.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
tests/drivers/clock_control: stm32h7_devices: Test clock api

Rework test to additionally test clock_on and clock_off APIs

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
drivers/clock_control: stm32h7: Add support for alt clocks

Add support for alternate clocks configuration.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
tests/drivers/clock_control: stm32h7_devices: Add TC using pll3_p

Add a test case using PLL3P output as SPI clock source.
Modify the test source to allow testing various tests sources.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
include/drivers: stm32_clock_control: Avoid non defined PLL outputs

On some series (H7, U5), it is possible define clock configuration
with disabled PLL outputs.
In that case, it is legit that matching pll property is not available.
Define corresponding STM32_PLLX_Y_DIVISOR macros using DT_PROP_OR
to avoid build issues in case prop is not available.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
tests/drivers/clock_control: stm32h7: device_test: Clean up

Cleanup the test a little bit to give some clarity.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
tests/drivers/clock_control: stm32h7_device: Use STM32_DT_CLOCKS_FOO

Make use of STM32_DT_CLOCKS_ macros to have the test work conditionally
based on alt clock presence.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
boards: nucleo_h723zg: Enable SPI

Enable SPI on nucleo_h723zg board

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
include/drivers/clock_control: stm32: Add _DT_CLOCK helpers

Add a DT based macros to be used by stm32 device drivers to
populate pclken[] arrays at build time.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
drivers/spi: stm32: Use alt clock freq if available

Add support for an alternate clock. If available,
alternate clock is enabled and used to get the
device clock rate.

Fixes #41650

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
ds/bindings/clocks: Add stm32 clock mux binding

Add STM32 clock mux binding.
Only property of a node using such compatible is to
select a clock input.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
drivers: clock_control: Add STM32 clock multiplexer driver

Add a clock multiplexer driver.
Its only function is to select a clock input.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
dts/arm/st: stm32h7: Add perck clock-mux node

Add perck clock-mux node as disabled by default and
the include bindings helping to select it.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
drivers/clock_control: stm32h7: Add support for CKPER clock mux

Add support for CKPER clock mux.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
tests/drivers/clock_control: stm32h7_device: Add test for CKPER source

Add 2 scenarios to test CKPER used as a clock source.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango commented:
Added clock mux support following discussion on L5 case [here](https://github.com/zephyrproject-rtos/zephyr/issues/42374)

FRASTM commented on drivers/clock_control/clock_stm32_ll_h7.c:
```diff
@@ -313,97 +313,43 @@ static inline int stm32_clock_control_on(const struct device *dev,
 	z_stm32_hsem_unlock(CFG_HW_RCC_SEMID);
 
-	return rc;
+	return 0;
```
is there any benefit to raise an error code here in case the DTS is malformed ?

erwango replied:
In a next commit, a check is added to verify that the requested clock is enabled.
Regarding the other parameters I'm not sure we can do much more checks.

FRASTM commented on tests/drivers/clock_control/stm32_clock_configuration/stm32h7_devices/boards/spi1_pll3p_1_d1ppre_4.overlay:
```diff
@@ -0,0 +1,31 @@
+/*
+ * Copyright (c) 2021 Linaro Limited
+ *
```
* Copyright (c) 202**2**

erwango committed:
include/drivers: clock_control.h: Add configure() to api

Add configure() function to clock_control API.
This function allows caller to configure a given clock.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango commented on include/drivers/clock_control.h:
```diff
@@ -150,6 +151,39 @@ static inline int clock_control_off(const struct device *dev,
+ * @param sys Opaque data representing the clock
+ * @return 0 on success, negative errno on failure.
+ */
+static inline int clock_control_select(const struct device *dev,
```
@nordic-krch Would you mind commenting this approval ?
I've added it on top of current PR, but I plan to raise it separately once we're aligned (and then rebase current PR on top)

erwango commented:
Rework to use new `clock_control_configure` API

erwango commented:
@galak, @gmarull I'd welcome your inputs if you have time.
